cf-execd/cf-execd-runagent.c

```diff
@@ -0,0 +1,150 @@
+/*
+  Copyright 2021 Northern.tech AS
+
+  This file is part of CFEngine 3 - written and maintained by Northern.tech AS.
+
+  This program is free software; you can redistribute it and/or modify it
+  under the terms of the GNU General Public License as published by the
+  Free Software Foundation; version 3.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA
+
+  To the extent this program is licensed as part of the Enterprise
+  versions of CFEngine, the applicable Commercial Open Source License
+  (COSL) may apply to this file if you as a licensee so wish it. See
+  included file COSL.txt.
+*/
+
+#include <stdio.h>
+
+#include <logging.h>
+#include <buffer.h>
+#include <json.h>
+#include <pipes.h>
+#include <alloc.h>
+#include <file_lib.h>
+#include <known_dirs.h>
+
+#include <cf-execd-runagent.h>
+
+
+/**
+ * Handle request to run cf-runagent.
+ *
+ * @note Expected to be called from a forked child process (blocks and forks).
+ */
+void HandleRunagentRequest(int conn_fd)
+{
+    FILE *conn_file = fdopen(conn_fd, "r+");
+    if (conn_file == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to create file stream from the connection file descriptor: %s",
+            GetErrorStr());
+
+        /* SIGH: There's no dprintf() on exotics. */
+        char error_buf[256];
+        int length = snprintf(error_buf, sizeof(error_buf),
+                              "{\"error\": \"Failed to create file stream from the connection file descriptor: %s\"}",
+                              GetErrorStr());
+        assert(length < sizeof(error_buf));
+        write(conn_fd, error_buf, MIN(sizeof(error_buf), (size_t) length));
```

Comment:
Could use an allocating printf function to avoid the hard coded buffer size limit.

Reply:
Yes, but this is a special code block because `fdopen()` is very unlikely to fail, one of the reasons could be problems with memory allocation. And the buffer is big enough for all `GetErrorStr()` strings, AFAICT.

Reply:
Makes sense 👍